Make wait() checks more reliable

Aparicio99 opened the pull request:
**Reasoning** 

I've caught an issue where pre/post rotate scripts were being reported with an error "error running non-shared prerotate script for...", but only on very specific but deterministic conditions. After some debugging I've found it to be due to on my use case logrotate being launched from crond which had already inherited from its parent to ignore SIGCHLD.

Even in this situation, the issue depended on compiler optimization flags, and even on moving/deleting non related code around, which eventually found to be due to the uninitialized use of the wait() return argument, therefore the changes I propose for every use of wait()/waitpid().

---
**Commit message**

If wait()/waitpid() returns an error for some reason it will not write to the *wstatus argument. Checking wstatus in this case is non deterministic due to the variable not being initialized, and may lead to wrongly report the forked command as executed successfully.

Also, one possible reason for wait() to return an error even if the forked command has executed successfully is the child process exiting before we call wait(). Normally when the child process finishes it stays in a zombie state until the wait() call, but if our parent process has set to ignore the SIGCHLD signal, we will inherited it and our child processes will exit immediately. Restoring the signal handler on initialization will avoid this issue.

cgzones commented on logrotate.c:
```diff
@@ -3292,6 +3292,11 @@ int main(int argc, const char **argv)
+    /* Restore SIGCHLD handler in case our parent process had it ignored.
+     * Not doing this may cause our forked childs to exit prematurely before our
+     * wait() call, making it impossible to known the child proper exit code */
+    signal(SIGCHLD, SIG_DFL);
```
Please issue a warning on failure, e.g.:
```c
if (signal(SIGCHLD, SIG_DFL) == SIG_ERR)
    message(MESS_WARN, "failed to reset SIGCHLD handler: %s\n", strerror(errno));
```

Aparicio99 replied:
Updated with your suggestion.

cgzones reviewed (changes requested):
Thanks, looks good to me except one small nit.

cgzones commented on logrotate.c:
```diff
@@ -545,8 +546,11 @@ static int runScript(const struct logInfo *log, const char *logfn, const char *l
 
-    wait(&rc);
-    return rc;
+    if (wait(&rc) > 0 && WIFEXITED(rc)) {
```
Could you please take this opportunity and replace with `wait(3)` with `waitpid(3)`.
```c
waitpid(pid, &rc, 0);
```

cgzones commented on logrotate.c:
```diff
@@ -753,9 +757,7 @@ static int shred_file(int fd, const char *filename, const struct logInfo *log)
-    wait(&status);
-
-    if (!WIFEXITED(status) || WEXITSTATUS(status)) {
+    if (wait(&status) < 0 || !WIFEXITED(status) || WEXITSTATUS(status)) {
```
```c
waitpid(pid, &status, 0);
```

cgzones commented on logrotate.c:
```diff
@@ -989,19 +991,19 @@ static int compressLogFile(const char *name, const struct logInfo *log, const st
-    fsync(outFile);
-
-    if (!WIFEXITED(status) || WEXITSTATUS(status)) {
+    if (wait(&status) < 0 || !WIFEXITED(status) || WEXITSTATUS(status)) {
```
```c
waitpid(pid, &status, 0);
```

cgzones reviewed (commented):
I'd like to take this opportunity and replace the `wait(3)` calls with `waitpid(3)`.
Also the `waitpid(3)` call on line 1112 is not checked.

Aparicio99 commented:
Replaced all wait() calls with waitpid() and added validation of all return codes.

strerror(errno) can also be added to the error messages, but since it would only apply to when waitpid() fails and not when "!WIFEXITED(status) || WEXITSTATUS(status)" is true, it would complicate the code and possible have 2 different error messages for each call. Not sure if it's worth it, but I could also add this if you think so.

cgzones commented:
> Replaced all wait() calls with waitpid() and added validation of all return codes.

Thanks.

> strerror(errno) can also be added to the error messages, but since it would only apply to when waitpid() fails and not when "!WIFEXITED(status) || WEXITSTATUS(status)" is true, it would complicate the code and possible have 2 different error messages for each call. Not sure if it's worth it, but I could also add this if you think so.

What about a helper like:

```c
int waitpid_checked(pid_t pid, const char **errmsg) {
    int r, status;

    r = waitpid(pid, &status, 0);
    if (r <= 0) {
        *errmsg = strerror(errno);
        return -1;
    }

    if (!WIFEXITED(status)) {
        *errmsg = "child did not terminate normally";
        return -1;
    }

    if (WEXITSTATUS(status) != 0) {
        *errmsg = "child did not exit with status 0";
        return -1;
    }

    *errmsg = NULL;
    return 0;
}
```
?

Aparicio99 commented:
Implemented the helper as you suggested. Also updated runScript and callers to use the helper error message.

cgzones commented:
Applied. Thanks for your contribution and the patch iterations.
